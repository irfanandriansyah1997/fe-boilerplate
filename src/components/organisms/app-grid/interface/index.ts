import { FC } from 'react';

import { ICellProps, IGridProps } from '../../../molecules/grid/interface';

/**
 * App Grid Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.30
 */
export interface IAppGridProps extends IGridProps {
  Cell: FC<ICellProps>;
  onForceRerender(): void;
  onUpdateGrid(): void;
}
