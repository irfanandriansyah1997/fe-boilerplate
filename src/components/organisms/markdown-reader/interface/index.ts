/**
 * Markdown Content Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMarkdownContentProps {
  content: string;
}

/**
 * Markdown Hooks Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMarkdownHooks {
  content?: string;
  status: number;
  url: string;
}

/**
 * Markdown Reader Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMarkdownReaderProps {
  markdownUrl: string;
}
