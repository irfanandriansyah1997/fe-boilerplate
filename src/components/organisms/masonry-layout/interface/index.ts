import { FC } from 'react';

/**
 * Grouping Masonry Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IGroupingMasonry {
  item: IMasonryItem[];
  key: string;
}

/**
 * Masonry Hooks Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.24
 */
export interface IMasonryHooks {
  gridCount: number;
  item: IGroupingMasonry[];
}

/**
 * Masonry Item Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMasonryItem {
  item: Record<string, any>;
  key: string;
}

/**
 * Masonry Layout Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMasonryLayout {
  component: FC<any>;
  item: IMasonryItem[];
  maxGrid: number;
}
