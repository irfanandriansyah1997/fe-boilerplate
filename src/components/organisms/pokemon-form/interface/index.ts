/**
 * Pokemon Form Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IPokemonFormProps {
  initialPokemonName?: string;
  onSubmit(pokemonName: string): void;
  pokemonName: string;
}

/**
 * Pokemon Option Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IPokemonOptionProps {
  onClick(pokemonName: string): void;
  pokemonName: string;
}
