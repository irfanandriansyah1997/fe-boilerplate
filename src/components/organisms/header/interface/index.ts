import { IMenuItem } from '../../../../interface/component';

/**
 * Header Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.14
 */
export interface IHeaderProps {
  menu: IMenuItem[];
}
