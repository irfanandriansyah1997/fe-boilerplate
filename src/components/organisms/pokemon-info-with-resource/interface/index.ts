import { IPokemonCachePayload } from '../../../../section/suspense-api/suspense-image/interface';
import { IPokemonCardTypeEnum } from '../../../molecules/pokemon-card/interface';

/**
 * Pokemon Info With Resource Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IPokemonInfoWithResourceProps {
  resource: IPokemonCachePayload;
  type?: IPokemonCardTypeEnum;
}
