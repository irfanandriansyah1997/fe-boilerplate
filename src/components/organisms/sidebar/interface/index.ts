import { IMenuItem } from '../../../../interface/component';

/**
 * Sidebar Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.14
 */
export interface ISidebarProps {
  menu: IMenuItem[];
  subtitle: string;
  title: string;
}
