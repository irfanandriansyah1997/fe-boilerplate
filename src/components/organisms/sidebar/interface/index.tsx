import { IMenu } from '../../../../interface/component';

/**
 * Sidebar Mobile Menu Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.24
 */
export interface ISidebarMobilMenuProps extends ISidebarProps {
  show: boolean;
  toggleShow(show: boolean): void;
}

/**
 * Sidebar Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.22
 */
export interface ISidebarProps extends IMenu {
  onClickBack(): void;
}
