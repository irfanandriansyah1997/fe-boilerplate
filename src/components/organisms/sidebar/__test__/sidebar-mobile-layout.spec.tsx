import { fireEvent, render } from '@testing-library/react';
import { ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';
import {
  TransitionProps,
  TransitionStatus
} from 'react-transition-group/Transition';

import { IMenuItem } from '../../../../interface/component';
import SidebarMobileLayout from '../section/mobile-layout.component';

const MOCK_MENU: IMenuItem[] = [
  {
    text: `Wrap Use Memo`,
    to: `/context/part-1`
  },
  {
    text: `Separate Context Value`,
    to: `/context/part-2`
  },
  {
    text: `Colocate State`,
    to: `/context/part-3`
  }
];

// Tell Jest to mock all timeout functions
jest.useFakeTimers();

jest.mock(`react-transition-group`, () => ({
  Transition: ({
    children,
    in: show,
    onExited
  }: TransitionProps<HTMLElement>) => {
    const fn = children as (status: TransitionStatus) => ReactNode;

    if (!show && onExited) onExited();

    return <div>{show ? fn(`entered`) : null}</div>;
  }
}));

describe(`Testing Sidebar On Mobile Site`, () => {
  describe(`Snapshot Testing`, () => {
    it(`Testing when at the start of rendering`, () => {
      const onClickBackSpy = jest.fn();
      const onClickItemSpy = jest.fn();
      const { getByTestId } = render(
        <SidebarMobileLayout
          onClickBack={onClickBackSpy}
          onClickItem={onClickItemSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      );

      expect(getByTestId(/^sidebar-msite$/i)).toBeInTheDocument();
      expect(getByTestId(/sidebar-msite-navbar-title/).textContent).toBe(
        `Title Course`
      );

      expect(getByTestId(/^sidebar-msite$/i)).toMatchSnapshot();
    });

    it(`Testing when user click toggle button`, () => {
      const onClickBackSpy = jest.fn();
      const onClickItemSpy = jest.fn();
      const { getByTestId, queryByTestId } = render(
        <MemoryRouter initialEntries={[`/context/part-1`]}>
          <SidebarMobileLayout
            onClickBack={onClickBackSpy}
            onClickItem={onClickItemSpy}
            subtitle="Sample Subtitle Course"
            title="Title Course"
            menu={MOCK_MENU}
          />
        </MemoryRouter>
      );

      expect(queryByTestId(/^sidebar-msite-dialog$/i)).not.toBeInTheDocument();
      expect(getByTestId(/^sidebar-msite$/i)).toBeInTheDocument();
      expect(getByTestId(/sidebar-msite-navbar-title/).textContent).toBe(
        `Title Course`
      );

      const button = getByTestId(/sidebar-msite-navbar-button/);

      fireEvent.click(button);
      expect(getByTestId(/^sidebar-msite-dialog$/i)).toBeInTheDocument();

      expect(getByTestId(/^sidebar-msite$/i)).toMatchSnapshot();
      expect(getByTestId(/^sidebar-msite-dialog$/i)).toMatchSnapshot();
    });
  });

  describe(`Testing simulate user scrolling browser`, () => {
    let addEventListenerSpy: jest.SpyInstance;
    let removeEventListenerSpy: jest.SpyInstance;

    beforeEach(() => {
      addEventListenerSpy = jest.spyOn(window, `addEventListener`);
      removeEventListenerSpy = jest.spyOn(window, `removeEventListener`);
    });

    it(`Testing component`, () => {
      const onClickBackSpy = jest.fn();
      const onClickItemSpy = jest.fn();
      const { getByTestId, unmount } = render(
        <MemoryRouter initialEntries={[`/context/part-1`]}>
          <SidebarMobileLayout
            onClickBack={onClickBackSpy}
            onClickItem={onClickItemSpy}
            subtitle="Sample Subtitle Course"
            title="Title Course"
            menu={MOCK_MENU}
          />
        </MemoryRouter>
      );

      /**
       * Check When First Render Is Called Add Event Listener
       */
      expect(window.addEventListener).toHaveBeenCalledWith(
        `scroll`,
        expect.any(Function)
      );

      expect(getByTestId(/^sidebar-msite$/i)).toBeInTheDocument();
      expect(getByTestId(/^sidebar-msite-navbar$/i)).toBeInTheDocument();
      expect(getByTestId(/^sidebar-msite-navbar$/i).className).not.toContain(
        `o-sidebar-mobile__navbar--scrolled`
      );

      fireEvent.scroll(window, { target: { scrollY: 30 } });
      expect(getByTestId(/^sidebar-msite-navbar$/i).className).toContain(
        `o-sidebar-mobile__navbar--scrolled`
      );

      /**
       * Simulate Unmount Component
       */
      unmount();
      expect(window.removeEventListener).toHaveBeenCalledWith(
        `scroll`,
        expect.any(Function)
      );
    });

    afterEach(() => {
      addEventListenerSpy.mockClear();
      removeEventListenerSpy.mockClear();
    });
  });

  it(`Simulate user click back to homepage button`, () => {
    const onClickBackSpy = jest.fn();
    const onClickItemSpy = jest.fn();
    const { getByTestId, getByText, queryByTestId } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <SidebarMobileLayout
          onClickBack={onClickBackSpy}
          onClickItem={onClickItemSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      </MemoryRouter>
    );

    expect(queryByTestId(/^sidebar-msite-dialog$/i)).not.toBeInTheDocument();

    const button = getByTestId(/sidebar-msite-navbar-button/);
    fireEvent.click(button);
    expect(getByTestId(/^sidebar-msite-dialog$/i)).toBeInTheDocument();

    const buttonBackToHomepage = getByText(/Back To Homepage/);
    fireEvent.click(buttonBackToHomepage);

    // Fast-forward time
    jest.runAllTimers();
    expect(onClickBackSpy).toHaveBeenCalled();
  });

  it(`Simulate user click back each of menu item`, async () => {
    const onClickBackSpy = jest.fn();
    const onClickItemSpy = jest.fn();
    const { findAllByTestId, getByTestId, queryByTestId } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <SidebarMobileLayout
          onClickBack={onClickBackSpy}
          onClickItem={onClickItemSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      </MemoryRouter>
    );

    expect(queryByTestId(/^sidebar-msite-dialog$/i)).not.toBeInTheDocument();

    const button = getByTestId(/sidebar-msite-navbar-button/);
    fireEvent.click(button);
    expect(getByTestId(/^sidebar-msite-dialog$/i)).toBeInTheDocument();

    const element = await findAllByTestId(/sidebar-msite-dialog-menu-item/);
    expect(element.length).toBe(3);

    /**
     * Simulate User Click Sidebar Item
     */
    fireEvent.click(element[1]);
    expect(onClickItemSpy).not.toHaveBeenCalledWith({});
  });

  it(`Simulate user click icon close on sidebar dialog`, async () => {
    const onClickBackSpy = jest.fn();
    const onClickItemSpy = jest.fn();
    const { getByTestId, getByText, queryByTestId } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <SidebarMobileLayout
          onClickBack={onClickBackSpy}
          onClickItem={onClickItemSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      </MemoryRouter>
    );

    expect(queryByTestId(/^sidebar-msite-dialog$/i)).not.toBeInTheDocument();

    const button = getByTestId(/sidebar-msite-navbar-button/);
    fireEvent.click(button);
    expect(getByTestId(/^sidebar-msite-dialog$/i)).toBeInTheDocument();

    const backButton = await getByText(/arrow_back/);
    fireEvent.click(backButton);

    expect(queryByTestId(/^sidebar-msite-dialog$/i)).not.toBeInTheDocument();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });
});
