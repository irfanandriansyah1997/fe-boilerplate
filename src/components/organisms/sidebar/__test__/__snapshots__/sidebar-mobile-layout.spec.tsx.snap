// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Sidebar On Mobile Site Snapshot Testing Testing when at the start of rendering 1`] = `
<div
  data-testid="sidebar-msite"
>
  <div
    class="o-sidebar-mobile__navbar"
    data-testid="sidebar-msite-navbar"
  >
    <button
      class="material-icons"
      data-testid="sidebar-msite-navbar-button"
      type="button"
    >
      menu
    </button>
    <h6
      data-testid="sidebar-msite-navbar-title"
    >
      Title Course
    </h6>
  </div>
</div>
`;

exports[`Testing Sidebar On Mobile Site Snapshot Testing Testing when user click toggle button 1`] = `
<div
  data-testid="sidebar-msite"
>
  <div
    class="o-sidebar-mobile__navbar"
    data-testid="sidebar-msite-navbar"
  >
    <button
      class="material-icons"
      data-testid="sidebar-msite-navbar-button"
      type="button"
    >
      menu
    </button>
    <h6
      data-testid="sidebar-msite-navbar-title"
    >
      Title Course
    </h6>
  </div>
</div>
`;

exports[`Testing Sidebar On Mobile Site Snapshot Testing Testing when user click toggle button 2`] = `
<div
  class="o-sidebar-mobile__dialog o-sidebar-mobile__dialog--entered"
  data-testid="sidebar-msite-dialog"
>
  <div
    class="o-sidebar-mobile__dialog-heading"
  >
    <h6>
      <button
        class="material-icons"
        type="button"
      >
        arrow_back
      </button>
      Title Course
    </h6>
    <p>
      Sample Subtitle Course
    </p>
  </div>
  <div
    class="o-sidebar__menu"
  >
    <a
      aria-current="page"
      class="o-sidebar__item o-sidebar__item--active"
      data-testid="sidebar-msite-dialog-menu-item"
      href="/context/part-1"
    >
      <span
        class="material-icons"
      >
        done
      </span>
      Wrap Use Memo
    </a>
    <a
      class="o-sidebar__item"
      data-testid="sidebar-msite-dialog-menu-item"
      href="/context/part-2"
    >
      <span
        class="material-icons"
      >
        done
      </span>
      Separate Context Value
    </a>
    <a
      class="o-sidebar__item"
      data-testid="sidebar-msite-dialog-menu-item"
      href="/context/part-3"
    >
      <span
        class="material-icons"
      >
        done
      </span>
      Colocate State
    </a>
  </div>
  <button
    class="o-sidebar-mobile__back-button"
    data-testid="sidebar-msite-dialog-click-back"
    type="button"
  >
    üè†¬†Back To Homepage
  </button>
</div>
`;
